### PR TITLE
[ADD] add module from la louve version 12 with the fixes.

### PR DESCRIPTION
**Pull request by houssine78**
https://gestion.coopiteasy.be/web#id=3503&view_type=form&model=project.task&action=479

### CODE REVIEW

File: pos_automatic_cashdrawer/README.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+*main Odoo server*. On the *POSbox*, you should install the module
+*hw_x* depending on the protocol implemented in your device.
+`Cashlogy <http://www.cashlogy.com>` are implemented in the
+*hw_cashlogy* module and also in pywebdriver <https://github.com/akretion/pywebdriver>.
+
+Configuration
+=============
+
+The cashlogyConnector adddress and port should be configured on the main Odoo server,
+in the menu *Pointof Sale > Configuration > Payment Methods*,under the *Point of Sale* tab.
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/README.rst:20]: Point of Sale > Configuration > Point of Sale > a POS : Automatic Cashdrawer checkbox

Improvement: Automatic Cashdrawer appears when IoT Box checkbox is checked.

File: pos_automatic_cashdrawer/views/pos_config.xml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+                        <label for="iface_automatic_cashdrawer" string="Automatic Cashdrawer"/>
+                        <div class="content-group row mt16" attrs="{'invisible': [('iface_automatic_cashdrawer', '=', False)]}">
+                            <label string="Automatic Cashdrawer IP address" for="iface_automatic_cashdrawer_ip_address" class="col-lg-3 o_light_label"/>
+                            <field name="iface_automatic_cashdrawer_ip_address"/>
+                            <label string="Automatic Cashdrawer TCP port" for="iface_automatic_cashdrawer_tcp_port" class="col-lg-3 o_light_label"/>
+                            <field name="iface_automatic_cashdrawer_tcp_port"/>
+                        </div>
+                    </div>
+                </div>
+            </xpath>
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/views/pos_config.xml:23]: It's for aesthetic only, but I would move this in the IoT Box checkbox: 
```xml
<xpath expr="//label[@for='iface_cashdrawer']/parent::div" position="after">
                <div id="iface_automatic_cashdrawer" class="row">
                    <label string="Automatic Cashdrawer"
                           for="iface_automatic_cashdrawer"
                           class="col-lg-3 o_light_label"/>
                    <field name="iface_automatic_cashdrawer"/>
                </div>
                <div class="content-group mt16" attrs="{'invisible': [('iface_automatic_cashdrawer', '=', False)]}">
                    <div>
                        <label string="IP address"
                               for="iface_automatic_cashdrawer_ip_address"
                               class="col-lg-3"/>
                        <field name="iface_automatic_cashdrawer_ip_address"/>
                    </div>
                    <div>
                        <label string="TCP port"
                               for="iface_automatic_cashdrawer_tcp_port"
                               class="col-lg-3"/>
                        <field name="iface_automatic_cashdrawer_tcp_port"/>
                    </div>
                </div>
</xpath>
```

File: pos_automatic_cashdrawer/static/src/js/chrome.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            It will check if the session has an opening balance cashbox and
+            if it doesn't, it will syncrhonize it as long as it's the same amount.
+            If it's different, it will ask for a manager and offer to overwrite it
+        */
+        init: function () {
+            var self = this;
+            this._super.apply(this, arguments);
+            this.ready.done(function () {
+                if (self.pos.config.iface_automatic_cashdrawer && self.pos.config.cash_control) {
+                	console.info("init");
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/chrome.js]: to remove

File: pos_automatic_cashdrawer/static/src/js/chrome.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        */
+        close: function () {
+            var self = this;
+            var args = arguments;
+            var _super = this._super;
+
+            if (!this.pos.config.iface_automatic_cashdrawer) {
+                return this._super.apply(this, arguments);
+            }
+            console.info("close");
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/chrome.js]: to remove

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            this._super.apply(this, arguments);
+            if (this.pos.config.iface_automatic_cashdrawer) {
+                this.message('cashlogy/connect', {
+                    'config': {
+                        'host': this.pos.config.iface_automatic_cashdrawer_ip_address,
+                        'port': this.pos.config.iface_automatic_cashdrawer_tcp_port
+                    }
+                });
+            }
+        },
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:36]: Unsure, but it seems an uncommon place to send the cashdrawer connection info. In standard, it only starts a loop that updates the connection status. [Here](https://github.com/OCA/pos/blob/12.0/pos_payment_terminal/static/src/js/devices.js#L50) a method is added to `devices.ProxyDevice` and called in [`screens.PaymentScreenWidget`](https://github.com/OCA/pos/blob/12.0/pos_payment_terminal/static/src/js/screens.js#L25) to send data

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+                'amount': amount,
+                'options': options,
+            });
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:65]: could be externalized in a single function

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_display_close_till: function () {
+            var self = this;
+            var done = this.message('cashlogy/display_close_till');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box'); 
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:88]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_display_complete_emptying: function () {
+            var self = this;
+            var done = this.message('cashlogy/display_complete_emptying');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:110]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_display_empty_stacker: function () {
+            var self = this;
+            var done = this.message('cashlogy/display_empty_stacker');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:132]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+                if (res !== amount) {
+                    console.error('Cashlogy', 'The dispensed amount was different than the requested', amount, res);
+                }
+            }).fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:165]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_start_add_change: function () {
+            var self = this;
+            var done = this.message('cashlogy/start_add_change');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:185]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_stop_acceptance: function () {
+            var self = this;
+            var done = this.message('cashlogy/stop_acceptance');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:216]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_get_inventory: function () {
+            var self = this;
+            var done = this.message('cashlogy/get_inventory');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:234]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+        automatic_cashdrawer_get_total_amount: function () {
+            var self = this;
+            var done = this.message('cashlogy/get_total_amount');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:252]: same here

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+        /*
+            Prints the current inventory of the machine
+        */
+        automatic_cashdrawer_print_inventory: function () {
+            this.pos.gui.show_popup('error', {
+                title: _t('Not implemented'),
+                body: _t('Method not implemented yet')
+            });
+            return $.Deferred().reject('Not implemented yet. TODO');
+        },
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:266]: Why use the function if not implemented ?

File: pos_automatic_cashdrawer/static/src/js/devices.js [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+            // }
+            var self = this;
+            var done = this.message('cashlogy/display_backoffice');
+            done.fail(function (error) {
+                var message = error ? error.data.message : _t('Cashdrawer not connected');
+                var body = error ? error.data.debug : _t('Make sure Cashdrawer connected with IOT Box');
+                self.pos.gui.show_popup('error-traceback', {
+                    'title': _t('Cashdrawer Error: ') + message,
+                    'body': body
+                });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/devices.js:287]: externalize in a single function to call when asynchronous function fails

File: pos_automatic_cashdrawer/static/src/js/models.js [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    ]);
+
+    models.load_fields("pos.config", [
+        'iface_automatic_cashdrawer',
+        'iface_automatic_cashdrawer_ip_address',
+        'iface_automatic_cashdrawer_tcp_port',
+        'group_pos_automatic_cashlogy_config',
+    ]);
+
+    // var Session = new Model('pos.session');
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/models.js]: remove

File: pos_automatic_cashdrawer/static/src/js/models.js [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+            });
+            return done;
+        },
+
+        // Sets the balance
+        action_set_balance: function (inventory, balance) {
+            var self = this;
+            console.info("action_set_balance");
+            console.info(balance);
+            console.info(inventory);
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/models.js]: remove if not useful to display in the console

File: pos_automatic_cashdrawer/static/src/js/screens.js [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+        TODO: Analyze why we need this here, on the numpad..
+        Isn't it already handled by the set_cashier function?
+    */
+    screens.NumpadWidget.include({
+        start: function () {
+            this._super();
+            this.gui.display_access_right_cashlogy(this.pos.get_cashier());
+        },
+    });
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/static/src/js/screens.js]: do we need to analyze?

File: pos_automatic_cashdrawer/views/assets.xml [hunk 20]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <template id="assets" inherit_id="point_of_sale.assets">
+        <xpath expr="." position="inside">
+            <script type="text/javascript" src="/pos_automatic_cashdrawer/static/src/js/models.js"></script>
+            <script type="text/javascript" src="/pos_automatic_cashdrawer/static/src/js/devices.js"></script>
+            <script type="text/javascript" src="/pos_automatic_cashdrawer/static/src/js/widgets.js"></script>
+            <script type="text/javascript" src="/pos_automatic_cashdrawer/static/src/js/screens.js"></script>
+            <script type="text/javascript" src="/pos_automatic_cashdrawer/static/src/js/chrome.js"></script>
```
Comment (vvrossem) [inline pos_automatic_cashdrawer/views/assets.xml]: > au final si tu n'a pas d'enfant dans ta balise c'est plus propre de la finir par />

**Review by vvrossem — CHANGES_REQUESTED** [review]
Quite big amount of code to review :hot_face:. I'd love to do a second one once hw_cashology is migrated and loaded on a posboxless.

File: pos_automatic_cashdrawer/README.rst (hunk 1 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/README.rst:20]: done !

File: pos_automatic_cashdrawer/views/pos_config.xml (hunk 2 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/views/pos_config.xml:23]: yeah but I try to stick to the original module. Ok to leave as it is  here ?

File: pos_automatic_cashdrawer/static/src/js/chrome.js (hunk 3 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/static/src/js/chrome.js]: done

File: pos_automatic_cashdrawer/static/src/js/chrome.js (hunk 4 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/static/src/js/chrome.js]: done

File: pos_automatic_cashdrawer/static/src/js/devices.js (hunk 15 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/static/src/js/devices.js:266]: I don't know ;)

File: pos_automatic_cashdrawer/static/src/js/models.js (hunk 17 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/static/src/js/models.js]: done

File: pos_automatic_cashdrawer/static/src/js/models.js (hunk 18 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/static/src/js/models.js]: done

File: pos_automatic_cashdrawer/static/src/js/screens.js (hunk 19 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/static/src/js/screens.js]: I won't analyze so I remove it

File: pos_automatic_cashdrawer/views/assets.xml (hunk 20 above)
Comment (houssine78) [reply pos_automatic_cashdrawer/views/assets.xml]: j'ai une fille qui a eu 6 ans hier, je fais quoi du coup ?

**Comment by houssine78** [comment]
@vvrossem I toke most of your comments into account but I've decided to leave some at it this to not diverge too much with the code of la louve as it will help to bring the future fixes and code improvement or better go to a common module in the futur.

**Comment by houssine78** [comment]
thanks for your review anyway

**Review by vvrossem — APPROVED** [review]

**Pull request merged**